sound/soc/sof/intel/hda.c

```diff
@@ -251,15 +261,56 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	}
 
 	/* codec detection */
-	if (!bus->codec_mask)
+	if (!bus->codec_mask) {
 		dev_info(bus->dev, "no hda codecs found!\n");
-	else
-		dev_info(bus->dev, "hda codecs found, mask %lx!\n", bus->codec_mask);
+	} else {
+		dev_info(bus->dev, "hda codecs found, mask %lx\n",
+			 bus->codec_mask);
+
+		for (i = 0; i < HDA_MAX_CODECS; i++) {
+			if (bus->codec_mask & (1 << i))
+				codec_num++;
+		}
+
+		/*
+		 * If no machine driver is found, then:
+		 *
+		 * hda machine driver is used if :
+		 * 1. there are HDMI codec and other HDA codecs
+		 * 2. only HDMI codec and NOCODEC is not enabled
+		 *
+		 * nocodec machine driver would be used if:
+		 * 1. only HDMI codec and but NOCODEC is enabled
+		 * 2. there is no codec found
+		 */
+		if (!pdata->machine) {
+			if (codec_num > 1)
+				hda_mach = snd_soc_acpi_intel_hda_machines;
+#if !IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
+			else if (HDMI_CODEC(bus->codec_mask))
+				hda_mach = snd_soc_acpi_intel_hda_machines;
+#endif
+
+			if (hda_mach) {
+				mach = pdata->desc->machines;
+				hda_mach->sof_fw_filename =
+					mach->sof_fw_filename;
+				pdata->machine = hda_mach;
+
+				dev_info(bus->dev, "using HDA machine driver %s now\n",
+					 hda_mach->drv_name);
+			}
+		}
+	}
 
 	/* used by hda machine driver to create dai links */
-	mach_params = (struct snd_soc_acpi_mach_params *)
-		&sdev->pdata->machine->mach_params;
-	mach_params->codec_mask = bus->codec_mask;
+	machine = sdev->pdata->machine;
+	if (machine) {
```

Comment:
no need for extra indirection, machine = pdata->machine would work (and be consistent with code above)

Reply:
yes, I will refine it. Thanks!